code/modules/vehicles/mecha/equipment/tools/work_tools.dm

```diff
@@ -158,67 +158,72 @@
+	create_reagents(400)
+	reagents.add_reagent(/datum/reagent/water, 400)
 
-/obj/item/mecha_parts/mecha_equipment/extinguisher/action(mob/source, atom/target, list/modifiers)
-	if(!action_checks(target) || get_dist(chassis, target)>3)
+/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/spray_extinguisher(mob/user)
+	if(reagents.total_volume < required_amount)
 		return
 
-	if(istype(target, /obj/structure/reagent_dispensers/watertank) && get_dist(chassis,target) <= 1)
-		var/obj/structure/reagent_dispensers/watertank/WT = target
-		WT.reagents.trans_to(src, 1000)
-		to_chat(source, "[icon2html(src, source)][span_notice("Extinguisher refilled.")]")
-		playsound(chassis, 'sound/effects/refill.ogg', 50, TRUE, -6)
-		return
+	var/list/turflist = list()
+	for(var/carddir in GLOB.cardinals)
+		turflist += get_step(chassis, carddir)
+		turflist += get_step(get_step(chassis, carddir), turn(carddir, 90))
 
-	if(reagents.total_volume <= 0)
-		return
-	playsound(chassis, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
+	for(var/turf/targetturf in turflist)
+		var/obj/effect/particle_effect/water/extinguisher/water = new /obj/effect/particle_effect/water/extinguisher(targetturf)
+		var/datum/reagents/water_reagents = new /datum/reagents(required_amount/8) //required_amount/8, because the water usage is split between eight sprays. As of this comment, required_amount/8 = 10u each.
+		water.reagents = water_reagents
+		water_reagents.my_atom = water
+		reagents.trans_to(water, required_amount/8)
+		water.move_at(get_step(chassis, get_dir(targetturf, chassis)), 2, 4) //Target is the tile opposite of the mech as the starting turf.
 
-	sprays_left += 5
-	add_hiddenprint(source) //log prints so admins can figure out who touched it last.
-	log_combat(source, target, "fired an extinguisher at")
-	spray_extinguisher(target)
-	return ..()
+	playsound(chassis, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
 
-/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/spray_extinguisher(atom/target)
-	var/direction = get_dir(chassis, target)
-	var/turf/T1 = get_turf(target)
-	var/turf/T2 = get_step(T1,turn(direction, 90))
-	var/turf/T3 = get_step(T1,turn(direction, -90))
-	var/list/targets = list(T1,T2,T3)
-
-	var/obj/effect/particle_effect/water/extinguisher/water = new /obj/effect/particle_effect/water/extinguisher(get_turf(chassis))
-	var/datum/reagents/water_reagents = new /datum/reagents(5)
-	water.reagents = water_reagents
-	water_reagents.my_atom = water
-	reagents.trans_to(water, 1)
-
-	var/delay = 2
-	var/datum/move_loop/our_loop = water.move_at(pick(targets), delay, 4)
-	RegisterSignal(our_loop, COMSIG_PARENT_QDELETING, .proc/water_finished_moving)
-
-/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/water_finished_moving(datum/move_loop/has_target/source)
-	SIGNAL_HANDLER
-	sprays_left--
-	if(!sprays_left)
+/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/attempt_refill(mob/usr)
+	if(reagents.maximum_volume == reagents.total_volume)
```

Comment:
(you should totally autodoc this proc btw)

Reply:
Also wtf is redefining usr in a proc like this allowed?
This should be user not usr imo

Reply:
Didn't even notice I used a different var, honestly.